### PR TITLE
Expose ConnectivityState of a ClientConn.

### LINKED ISSUE #1208: channel state changes
"All libraries should also expose an API that enables the application (user of the gRPC API) to be notified when the channel state changes." (https://github.com/grpc/grpc/blob/master/doc/connectivity-semantics-and-api.md)

Has this been implemented yet...? I can't see any way of getting state changes on the client...

### PR DESCRIPTION
**Pull request by MakMukhi**
fixes #1208
#1359

### CODE REVIEW

File: clientconn.go [hunk 1]
```diff
@@ -476,6 +478,97 @@ func (s ConnectivityState) String() string {
 	}
 }
 
+// connectivityStateEvaluator gets updated by addrConns when their
+// stats transition, based on which it evaluates the state of
```
Comment (dfawley) [inline clientconn.go]: *states

File: clientconn.go [hunk 2]
```diff
@@ -476,6 +478,97 @@ func (s ConnectivityState) String() string {
...
+// ClientConn.
+// Note: This code will eventully sit in the blanacer in the new design.
```
Comment (dfawley) [inline clientconn.go]: *eventually

File: clientconn.go [hunk 3]
```diff
@@ -476,6 +478,97 @@ func (s ConnectivityState) String() string {
...
+}
+
+// recordTransition records state change happening in every addrConn and based on
+// that it evaluates what state the ClientConn is in.
+// It can only transition between Ready, Connecting and TransientFailure. Other states,
+// Idle and Shutdown are transitioned into by ClientConn; in the begining of the conneciton
+// before any addrConn is created ClientConn is in idle state. In the end when ClientConn
+// closes it is in Shutdown state.
+// TODO Note that in later releases, a ClientConn with no activity will be put into an Idle state.
+func (csEvltr *connectivityStateEvaluator) recordTransition(oldState, newState ConnectivityState) {
```
Comment (dfawley) [inline clientconn.go]: csEvltr -> c?  cse?

File: clientconn.go [hunk 4]
```diff
@@ -841,8 +923,9 @@ func (ac *addrConn) resetTransport(closeTransport bool) error {
 			ac.down(downErrorf(false, true, "%v", errNetworkIO))
 			ac.down = nil
 		}
+		oldState := ac.state
```
Comment (dfawley) [inline clientconn.go]: How about this:

```go
updateState := func(new ConnectivityState) {
  ac.csEvltr.recordTransition(ac.state, new)
  ac.state = new
}
....
updateState(Connecting)
...
updateState(TransientFailure)
...
```

File: clientconn_test.go [hunk 5]
```diff
@@ -27,10 +28,60 @@ import (
...
+	}()
+
+	cc, err := Dial("foo.bar.com", WithBalancer(RoundRobin(resolver)), WithInsecure())
+	if err != nil {
+		t.Fatalf("Dial(\"foo.bar.com\", WithBalancer(_)) = _, %v, want _ <nil>", err)
+	}
+	defer cc.Close()
+	wantState := Ready
+	if state, ok := assertState(wantState, cc); !ok {
+		t.Fatalf("asserState(%s) = %s, false, want %s, true", wantState, state, wantState)
```
Comment (dfawley) [inline clientconn_test.go]: *assertState

File: clientconn.go [hunk 6]
```diff
@@ -476,6 +478,97 @@ func (s ConnectivityState) String() string {
 	}
 }
 
+// connectivityStateEvaluator gets updated by addrConns when their
+// states transition, based on which it evaluates the state of
+// ClientConn.
+// Note: This code will eventually sit in the blanacer in the new design.
```
Comment (dzbarsky) [inline clientconn.go]: nit: balancer is typo'ed

File: clientconn.go [hunk 7]
```diff
@@ -476,6 +478,97 @@ func (s ConnectivityState) String() string {
...
+	mu                  sync.Mutex
+	numReady            uint64 // Number of addrConns in ready state.
+	numConnecting       uint64 // Number of addrConns in connecting state.
+	numTransientFailure uint64 // Number of addrConns in transientFailure.
+}
+
+// recordTransition records state change happening in every addrConn and based on
+// that it evaluates what state the ClientConn is in.
+// It can only transition between Ready, Connecting and TransientFailure. Other states,
+// Idle and Shutdown are transitioned into by ClientConn; in the begining of the conneciton
```
Comment (dzbarsky) [inline clientconn.go]: nit: begining, conneciton

File: clientconn.go [hunk 8]
```diff
@@ -476,6 +478,97 @@ func (s ConnectivityState) String() string {
...
+// connectivityStateManager keeps the ConnectivityState of ClientConn.
+// This struct will eventually be exported so the balancers can access it.
+type connectivityStateManager struct {
+	mu         sync.Mutex
+	state      ConnectivityState
+	notifyChan chan struct{}
+}
+
+// updateState updates the ConnectivityState of ClientConn.
+// If there's a change it notifies goroutines waiting on state change to
```
Comment (dzbarsky) [inline clientconn.go]: it can only notify 1 goroutine, right?

File: clientconn.go [hunk 9]
```diff
@@ -522,14 +634,18 @@ func (cc *ClientConn) lbWatcher(doneChan chan struct{}) {
...
 				if err == nil {
 					close(doneChan)
 					doneChan = nil
 				}
 			} else {
-				cc.resetAddrConn(a, false, nil)
+				err = cc.resetAddrConn(a, false, nil)
+			}
+			if err != nil {
+				grpclog.Warningf("Error creating conneciton to %v. Err: %v", a, err)
```
Comment (dzbarsky) [inline clientconn.go]: nit: connection

File: clientconn.go (hunk 8 above)
Comment (MakMukhi) [inline clientconn.go]: Although rare, but it's possible multiple goroutines can make a call to WaitForStateChange

**Pull request merged**